File: packages/js-dpp/lib/stateTransition/fee/operations/DeleteOperation.js
const AbstractOperation = require('./AbstractOperation');

const {
  DELETE_BASE_PROCESSING_COST,
  PROCESSING_CREDIT_PER_BYTE,
  STORAGE_CREDIT_PER_BYTE,
} = require('../constants');

class DeleteOperation extends AbstractOperation {
  /**
   * @param {number} keySize
   * @param {number} valueSize
   */
  constructor(keySize, valueSize) {
    super();

    this.keySize = keySize;
    this.valueSize = valueSize;
  }

  /**
   * Get CPU cost of the operation
   *
   * @returns {number}
   */
  getProcessingCost() {
    return ((this.keySize + this.valueSize) * PROCESSING_CREDIT_PER_BYTE)
      + DELETE_BASE_PROCESSING_COST;
  }

  /**
   * Get storage cost of the operation
   *
   * @returns {number}
   */
  getStorageCost() {
    return -((this.keySize + this.valueSize) * STORAGE_CREDIT_PER_BYTE);
  }

  /**
   * @return {{keySize: number, type: string, valueSize: number}}
   */
  toJSON() {
    return {
      type: 'delete',
      keySize: this.keySize,
      valueSize: this.valueSize,
    };
  }

  /**
   * @param {{keySize: number, type: string, valueSize: number}} json
   * @return {DeleteOperation}
   */
  static fromJSON(json) {
    return new DeleteOperation(json.keySize, json.valueSize);
  }
}

module.exports = DeleteOperation;

File: packages/rs-dpp/src/state_transition/fee/operations/delete_operation.rs
use serde::{Deserialize, Serialize};

use super::OperationLike;
use crate::{
    prelude::Fee,
    state_transition::fee::constants::{
        DELETE_BASE_PROCESSING_COST, PROCESSING_CREDIT_PER_BYTE, STORAGE_CREDIT_PER_BYTE,
    },
};

#[derive(Default, Debug, Clone, Copy, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub struct DeleteOperation {
    pub key_size: usize,
    pub value_size: usize,
}

impl DeleteOperation {
    pub fn new(key_size: usize, value_size: usize) -> Self {
        Self {
            key_size,
            value_size,
        }
    }
}

impl OperationLike for DeleteOperation {
    fn get_processing_cost(&self) -> Fee {
        let current_cost = (self.key_size.saturating_add(self.value_size) as Fee)
            .saturating_mul(PROCESSING_CREDIT_PER_BYTE);

        DELETE_BASE_PROCESSING_COST.saturating_add(current_cost)
    }

    fn get_storage_cost(&self) -> Fee {
        -(self.key_size.saturating_add(self.value_size) as Fee)
            .saturating_mul(STORAGE_CREDIT_PER_BYTE)
    }
}
